File: mdx-js/level3.js
/// class VertexPosition
/// Desc A vector describing a 3D vertex
function MdxVertexPosition(stream) {
	this.position = stream.readFloat32Array(3);
}

/// class VertexNormal
/// Desc A vector describing a 3D normal
function MdxVertexNormal(stream) {
	this.normal = stream.readFloat32Array(3);
}

/// class FaceTypeGroup
/// Desc ?
function MdxFaceTypeGroup(stream) {
	this.faceType = stream.readUint32();
}

/// class FaceGroup
/// Desc ?
function MdxFaceGroup(stream) {
	this.nrOfIndexes = stream.readUint32();
}

/// class Face
/// Desc A triplet of integers that describe indices to vertices that form a triangle
function MdxFace(stream) {
	this.index1 = stream.readUint16();
	this.index2 = stream.readUint16();
	this.index3 = stream.readUint16();
}

/// class VertexGroup
/// Desc ?
function MdxVertexGroup(stream) {
	this.matrixGroup = stream.readUint8();
}

/// class MatrixGroup
/// Desc ?
function MdxMatrixGroup(stream) {
	this.matrixGroupSize = stream.readUint32();
}

/// class MatrixIndex
/// Desc ?
function MdxMatrixIndex(stream) {
	this.matrixIndex = stream.readUint32();
}

/// class Extent
/// Desc ?
function MdxExtent(stream) {
	this.boundsRadius = stream.readFloat32();
	this.minimumExtent = stream.readFloat32Array(3);
	this.maximumExtent = stream.readFloat32Array(3);
}

/// class VertexTexturePosition
/// Desc A vector describing a 2D point in a texture
function MdxVertexTexturePosition(stream) {
	this.texturePosition = stream.readFloat32Array(2);
}

File: parser/level3.rb
class VertexPosition
	attr_accessor :position
	
	def initialize (stream)
		@position = stream.readVector(3)
	end
	
	def to_s ()
		return "[%.2f, %.2f, %.2f]" % @position
	end
end

class VertexNormal
	attr_accessor :normal
	
	def initialize (stream)
		@normal = stream.readVector(3)
	end
	
	def to_s ()
		return "[%.2f, %.2f, %.2f]" % @normal
	end
end

class FaceTypeGroup
	attr_accessor :faceType
	
	def initialize (stream)
		@faceType = stream.readUInt()
	end
	
	def to_s ()
		return @faceType.to_s
	end
end

class FaceGroup
	attr_accessor :nrOfIndexes
	
	def initialize (stream)
		@nrOfIndexes = stream.readUInt()
	end
	
	def to_s ()
		return @nrOfIndexes.to_s
	end
end

class Face
	attr_accessor :index1, :index2, :index3
	
	def initialize (stream)
		@index1 = stream.readUShort()
		@index2 = stream.readUShort()
		@index3 = stream.readUShort()
	end
	
	def to_s ()
		return "[%d, %d, %d]" % [@index1, @index2, @index3]
	end
end

class VertexGroup
	attr_accessor :matrixGroup
	
	def initialize (stream)
		@matrixGroup = stream.readUByte()
	end
	
	def to_s ()
		return @matrixGroup.to_s
	end
end

class MatrixGroup
	attr_accessor :matrixGroupSize
	
	def initialize (stream)
		@matrixGroupSize = stream.readUInt()
	end
	
	def to_s ()
		return @matrixGroupSize.to_s
	end
end

class MatrixIndex
	attr_accessor :matrixIndex
	
	def initialize (stream)
		@matrixIndex = stream.readUInt()
	end
	
	def to_s ()
		return @matrixIndex.to_s
	end
end

class Extent
	attr_accessor :boundsRadius, :minimumExtent, :maximumExtent
	
	def initialize (stream)
		@boundsRadius = stream.readFloat()
		@minimumExtent = stream.readVector(3)
		@maximumExtent = stream.readVector(3)
	end
	
	def to_s ()
		return "[%.2f, [%.2f, %.2f, %.2f], [%.2f, %.2f, %.2f]]" % [@boundsRadius, @minimumExtent, @maximumExtent]
	end
end

class VertexTexturePosition
	attr_accessor :texturePosition
	
	def initialize (stream)
		@texturePosition = stream.readVector(2)
	end
	
	def to_s ()
		return "[%.2f, %.2f]" % @texturePosition
	end
end

class Vertex
	attr_accessor :position
	
	def initialize (stream)
		@position = stream.readVector(3)
	end
	
	def to_s ()
		return "[%.2f, %.2f, %.2f]" % @position
	end
end
